File: spoj/classicals/DIV.cpp
#include <cstdio>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

#define ll long long

ll isprime[1000001];

ll pow(ll a, ll b){
	if (b ==1) return a;
	if(b%2){
		return a*pow(a, b-1);
	}
	ll x = pow(a, b/2);
	return x*x;
}

vector<ll> sieve(){
	fill_n(isprime, 1000001, 1);
	isprime[0] = isprime[1] = 0;
	for (ll i =2 ; i < 1001; i += 1){
		if(isprime[i]){
			for (ll j = i*i; j < 1000001; j += i){
				isprime[j] = 0;
			}
		}
	}
	vector<ll> temp;
	for (ll i = 2; i < 1000001; i += 1){
		if(isprime[i]){
			temp.push_back(i);
		}
	}
	return temp;
}

int main(void){

	vector<ll> primes = sieve();
	// cout << "Here are some primes: " << endl;
	// for (ll i = 0; i< 20; i += 1){
	// 	cout << primes[i] << " ";
	// }
	// cout << endl;
	vector<ll> ans;
	ll P = primes.size();
	cout << "Total primes are: " << P << endl;
	ll limit = 1000001;
	ll x, y;
	for (ll i = 0; i < P-1; i += 1){
		for (ll j = i+1; j < P; j += 1){
			x = 0; 
			y = 0;
			if(primes[i]*primes[j] >= limit) break;
			// cout << "primes[i]= " << primes[i] << " primes[j]= " << primes[j] << endl;
			while (pow(primes[i], primes[x]) < limit){
				y = 0;
				while(pow(primes[i], primes[x]-1)*pow(primes[j], primes[y]-1) < limit){
					// cout << "primes[i]= " << primes[i] << " primes[j]= " << primes[j] << " primes[x]= " << primes[x] << " primes[y] " << primes[y] << " pow1: " << pow(primes[i], primes[x]-1) << " pow2: " << pow(primes[j], primes[y]-1) << endl;
					if(primes[x] != primes[y]){
						// cout << "Pushing: " << pow(primes[i], primes[x]-1)*pow(primes[j], primes[y]-1) << endl;
						ans.push_back(pow(primes[i], primes[x]-1)*pow(primes[j], primes[y]-1));
					}
					y += 1;
				}
				x += 1;
			}
		}
	}
	ll rems[] = {6, 10, 14, 22, 15, 21};
	ll bases[] = {2, 3, 5, 7};
	for (ll i = 0; i < 4; i += 1){
		for (ll j = 0; j < 6; j += 1){
			ll x = pow(bases[i], rems[j]-1);
			if(x > 0 && x < limit){
				ans.push_back(x);
			} 
		}
	}
	ans.push_back(pow(11, 5));
	sort(ans.begin(), ans.end());
	for (ll i = 8; i < ans.size(); i += 9){
		cout << ans[i] << endl;
	}
}

File: spoj/classicals/DIV.py
limit = 10**6 + 1

isprime = [1]*limit

isprime[0] = isprime[1] = 0
for i in xrange(4,limit,2):
	isprime[i] = 0
for i in xrange(3,int(limit**0.5)+1,2):
	if(isprime[i]):
		for j in xrange(i*i, limit, 2*i):
			isprime[j] = 0

primes = [2]

for i in xrange(3, limit, 2):
	if(isprime[i]):
		primes.append(i)

ans = []
for i in xrange(len(primes)-1):
	s = primes[i]
	for j in xrange(i+1, len(primes)):
		t = primes[j]
		if t == s:
			continue
		if s*pow(t,2) > limit and t*pow(s,2) > limit:
			break
		for u in primes:
			tx = pow(s,u-1)
			if tx > limit:
				break
			for v in primes:
				if u == v: continue
				ty = pow(t,v-1)
				if ty>limit or tx*ty > limit:
					break
				else:
					if(tx*ty == 999952): print s, t, u, v
					ans.append(tx*ty)

powers = [6, 10, 14, 15, 21]
bases = [2, 3, 5, 7]
for base in bases:
	for power in powers:
		if pow(base, power-1) < limit:
			ans.append(pow(base, power-1))

ans.sort()
tocheck = [50, 99, 162, 999524, 999728, 999927]
for s in tocheck:
	print ans.index(s), (ans.index(s)+1)%9